ca/certificate-authority.go

```diff
@@ -215,176 +217,397 @@ func (ca *CertificateAuthorityImpl) RevokeCertificate(serial string, reasonCode
+func uniqueHostNames(csr *x509.CertificateRequest) (commonName string, hostNames []string) {
+	hostNames = make([]string, len(csr.DNSNames))
+	copy(hostNames, csr.DNSNames)
+	if len(csr.Subject.CommonName) > 0 {
+		hostNames = append(hostNames, csr.Subject.CommonName)
+	}
+
+	hostNames = core.UniqueNames(hostNames)
+
+	if len(csr.Subject.CommonName) > 0 {
+		commonName = csr.Subject.CommonName
+	} else if len(hostNames) > 0 {
+		commonName = hostNames[0]
+	}
+	return
+}
+
+// NewCertificateRequest vets a CSR, and if it's one that we're willing to
+// issue under, creates a CertificateRequest so that it can be used for
+// issuance later.
+func (ca *CertificateAuthorityImpl) NewCertificateRequest(req core.CertificateRequest) (core.CertificateRequest, error) {
+	emptyCertRequest := core.CertificateRequest{}
+
+	// Parse the CSR
+	csr, err := x509.ParseCertificateRequest(req.CSR)
+	if err != nil {
+		return emptyCertRequest, err
+	}
+
+	// Verify that the public key in the CSR is acceptable
 	key, ok := csr.PublicKey.(crypto.PublicKey)
 	if !ok {
 		err = core.MalformedRequestError("Invalid public key in CSR.")
 		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-		ca.log.AuditErr(err)
-		return emptyCert, err
+		ca.Log.AuditErr(err)
+		return emptyCertRequest, err
 	}
 	if err = core.GoodKey(key); err != nil {
 		err = core.MalformedRequestError(fmt.Sprintf("Invalid public key in CSR: %s", err.Error()))
 		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-		ca.log.AuditErr(err)
-		return emptyCert, err
+		ca.Log.AuditErr(err)
+		return emptyCertRequest, err
 	}
 	if badSignatureAlgorithms[csr.SignatureAlgorithm] {
 		err = core.MalformedRequestError("Invalid signature algorithm in CSR")
 		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-		ca.log.AuditErr(err)
-		return emptyCert, err
+		ca.Log.AuditErr(err)
+		return emptyCertRequest, err
 	}
 
+	// Verify that the hostnames in the CSR are ones that we will issue for
 	// Pull hostnames from CSR
 	// Authorization is checked by the RA
-	commonName := ""
-	hostNames := make([]string, len(csr.DNSNames))
-	copy(hostNames, csr.DNSNames)
-	if len(csr.Subject.CommonName) > 0 {
-		commonName = csr.Subject.CommonName
-		hostNames = append(hostNames, csr.Subject.CommonName)
-	} else if len(hostNames) > 0 {
-		commonName = hostNames[0]
-	} else {
+	commonName, hostNames := uniqueHostNames(csr)
+	if len(hostNames) == 0 {
 		err = core.MalformedRequestError("Cannot issue a certificate without a hostname.")
 		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-		ca.log.AuditErr(err)
-		return emptyCert, err
+		ca.Log.AuditErr(err)
+		return emptyCertRequest, err
 	}
 
-	// Collapse any duplicate names.  Note that this operation may re-order the names
-	hostNames = core.UniqueNames(hostNames)
+	// Collapse any duplicate names and check that there aren't too many.
+	// Note that this operation may re-order the names.
```

Comment:
This comment isn't relevant any more.  We're just checking that there aren't too many names.